azure/services/routetables/routetables.go

```diff
@@ -44,78 +47,75 @@ type Service struct {
 }
 
 // New creates a new service.
-func New(scope *scope.ClusterScope) *Service {
+func New(scope RouteTableScope) *Service {
 	return &Service{
 		Scope:  scope,
 		client: newClient(scope),
 	}
 }
 
-// Reconcile gets/creates/updates a route table.
+// Reconcile gets/creates/updates route tables.
 func (s *Service) Reconcile(ctx context.Context) error {
 	ctx, span := tele.Tracer().Start(ctx, "routetables.Service.Reconcile")
 	defer span.End()
 
-	if !s.Scope.Vnet().IsManaged(s.Scope.ClusterName()) {
+	ctx, cancel := context.WithTimeout(ctx, reconciler.DefaultAzureServiceReconcileTimeout)
+	defer cancel()
+
+	managed, err := s.Scope.IsVnetManaged()
+	if err != nil {
+		return errors.Wrap(err, "failed to check if vnet is managed")
+	} else if !managed {
 		s.Scope.V(4).Info("Skipping route tables reconcile in custom vnet mode")
 		return nil
 	}
 
-	for _, routeTableSpec := range s.Scope.RouteTableSpecs() {
-		existingRouteTable, err := s.Get(ctx, s.Scope.ResourceGroup(), routeTableSpec.Name)
-		if !azure.ResourceNotFound(err) {
-			if err != nil {
-				return errors.Wrapf(err, "failed to get route table %s in %s", routeTableSpec.Name, s.Scope.ResourceGroup())
-			}
-
-			// route table already exists
-			// currently don't support specifying your own routes via spec
-			routeTableSpec.Subnet.RouteTable.Name = to.String(existingRouteTable.Name)
-			routeTableSpec.Subnet.RouteTable.ID = to.String(existingRouteTable.ID)
-			s.Scope.SetSubnet(routeTableSpec.Subnet)
+	// We go through the list of route tables to reconcile each one, independently of the result of the previous one.
+	// If multiple erros occur, we return the most pressing one
+	// order of precedence is: error creating -> creating in progress -> created (no error)
+	var result error
 
-			continue
-		}
-
-		s.Scope.V(2).Info("creating Route Table", "route table", routeTableSpec.Name)
-		err = s.client.CreateOrUpdate(
-			ctx,
-			s.Scope.ResourceGroup(),
-			routeTableSpec.Name,
-			network.RouteTable{
-				Location:                   to.StringPtr(s.Scope.Location()),
-				RouteTablePropertiesFormat: &network.RouteTablePropertiesFormat{},
-			},
-		)
-		if err != nil {
-			return errors.Wrapf(err, "failed to create route table %s in resource group %s", routeTableSpec.Name, s.Scope.ResourceGroup())
+	for _, rtSpec := range s.Scope.RouteTableSpecs() {
+		if err := async.CreateResource(ctx, s.Scope, s.client, rtSpec, serviceName); err != nil {
+			if !azure.IsOperationNotDoneError(err) || result == nil {
+				result = err
+			}
 		}
-		s.Scope.V(2).Info("successfully created route table", "route table", routeTableSpec.Name)
 	}
```

Comment:
I have the same thought as https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/1684#issuecomment-931091181 , https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/1684#discussion_r719205970 about multiple errors case especially multiple non operation-not-done errors. And also about tests for the current implementation